apps/admin/backend/src/app.ts

```diff
@@ -325,35 +333,61 @@ function buildApi({
       return store.getSystemSettings(electionId);
     },
 
-    // `configure` and `unconfigure` handle changes to the election definition
-    async configure(input: {
-      electionData: string;
-      systemSettingsData: string;
-    }): Promise<Result<{ electionId: Id }, ConfigureError>> {
-      const electionDefinitionParseResult = safeParseElectionDefinition(
-        input.electionData
-      );
-      if (electionDefinitionParseResult.isErr()) {
-        return err({
-          type: 'invalidElection',
-          message: electionDefinitionParseResult.err().message,
-        });
+    async listPotentialElectionPackagesOnUsbDrive(): Promise<
+      Result<FileSystemEntry[], ListDirectoryOnUsbDriveError>
+    > {
+      const usbDriveEntriesResult = await listDirectoryOnUsbDrive(usbDrive, '');
+      if (usbDriveEntriesResult.isErr()) {
+        return usbDriveEntriesResult;
       }
-      const electionDefinition = electionDefinitionParseResult.ok();
 
-      const systemSettingsParseResult = safeParseSystemSettings(
-        input.systemSettingsData
+      return ok(
+        usbDriveEntriesResult
+          .ok()
+          .filter(
+            (entry) =>
+              entry.type === FileSystemEntryType.File &&
+              entry.name.endsWith('.zip')
+          )
+          // Most recent first
+          .sort((a, b) => b.ctime.getTime() - a.ctime.getTime())
       );
-      if (systemSettingsParseResult.isErr()) {
-        return err({
-          type: 'invalidSystemSettings',
-          message: systemSettingsParseResult.err().message,
-        });
+    },
+
+    // `configure` and `unconfigure` handle changes to the election definition
+    async configure(input: {
+      electionFilePath: string;
+    }): Promise<Result<{ electionId: Id }, ElectionPackageError>> {
+      let electionPackage: ElectionPackage;
+
+      if (input.electionFilePath.endsWith('.json')) {
+        const electionDefinitionResult = safeParseElectionDefinition(
+          await fs.readFile(input.electionFilePath, 'utf8')
```

Comment:
This reads a file without doing any validation that the path is where we expect it to be. Perhaps the API should concern itself only with filenames and not full paths?

Reply:
Where do we expect those files to be? Perhaps we only allow files from mounted USB drives in prod?

Reply:
That seems like a reasonable constraint to me. @mattroe what do you think?

Reply:
Do you mean in the new UI or in the manual old file picker? I think the old file picker should support any path.

Reply:
The problem with the way it is now is that a compromised frontend can make the backend read any file on the system and load it into memory. This might also include `/dev/` files that would result in a crash of the system or potentially alter system state. For example, passing `/dev/urandom` would cause the backend to read for quite a while until all available memory and swap is used up, making the system less responsive and possibly crashing the backend and/or the frontend.

Personally, I think allowing any path in dev (or with some other flag) and restricting it to mounted USB drives in prod is a reasonable compromise. If we decide to leave this as-is for now I strongly recommend creating an issue to follow up with a fix later.

Reply:
I'm confused was the reading of the file already on the backend before this change? Is it possible to not change where that is at least for the manual file picker for the purposes of the release? I'm also not understanding if you just mean filenames or relative paths to the root directory @eventualbuddha , we absolutely need to be able to load things from subdirectories on the usb. We do not need to load things outside of a mounted usb in prod, the file permissions in kiosk-browser have made that impossible in prod to date but is that not relevant if we're reading from the backend? We definitely can not introduce the ability to do that with this change.

Reply:
Agreed, we need to limit location of files read to USB in prod.

Reply:
Picking this up...

If my understanding of this thread is correct, I think we just want to add a prod-specific guard to the backend method, e.g. in the form of an `assert`, as a defense-in-depth measure. Sounds like the kiosk-browser file picker is already configured to only display removable media in prod (curious to find where exactly this logic lives).

Reply:
^ Will be sure to use a proper glob matching library, like we do in `Exporter`, so that we don't mistakenly allow a path like `/media/../dev`

Reply:
It displays non-removable media but today when you select it you get a permissions reading error about it, but I'm not sure where the guard for that lives and if moving things ot the backend here changes that

Reply:
https://github.com/votingworks/vxsuite/pull/4433/commits/6935fe154a187074c1a0cf260a511aac5ceed7da